src/rgw/rgw_sal.h

```diff
@@ -476,6 +479,13 @@ class User {
     User(User& _o) = default;
     virtual ~User() = default;
 
+
+    /*returns whether info is empty, as part of making Bucket fully abstract -Daniel P*/
+    virtual bool info_empty()
+    {
+      return info.user_id.id.empty();
```

Comment:
ok, so the problem with `if (rgw::sal::User::empty(user))` is that it's not polymorphic:
```c++
static bool empty(User* u) { return (!u || u->info.user_id.id.empty()); }
static bool empty(std::unique_ptr<User>& u) { return (!u || u->info.user_id.id.empty()); }
```
can these `empty()` functions just use your virtual `info_empty()` function instead?
```c++
static bool empty(User* u) { return (!u || u->info_empty()); }
static bool empty(std::unique_ptr<User>& u) { return (!u || u->info_empty()); }
```
that way you don't have to change every caller of `rgw::sal::User::empty()`

p.s. for @dang, is there really a use-case for a default-constructed `User`? can't we just make it a precondition, that you don't construct a `User` until you have a uid for it? that way we don't have this extra uninitialized state, and don't need `info_empty()` at all

Reply:
good idea